File: services/workers/src/workers/splitter.py
"""Stage 1 — Splitter worker (PRD §6.2).

Cheap, CPU-only, never touches Docling. Downloads the source once,
extracts the page count + outline, writes shard rows, and fans out one
doc.parse job per shard.
"""

from __future__ import annotations

import tempfile
from pathlib import Path

from core.config import get_settings
from core.db import repo
from core.db.models import DocState
from core.errors import ErrorCode, PlatformError
from core.queue import contracts, streams
from core.storage import s3
from parsing.splitter import chapter_aligned_bounds, extract_outline, fixed_bounds


def handle_split(session, job: dict, redis) -> None:
    settings = get_settings()

    doc_id = job["doc_id"]
    doc = repo.get_document(session, doc_id)
    if doc is None:
        raise PlatformError(ErrorCode.PDF_CORRUPT, f"document {doc_id} vanished")

    # Idempotency: re-delivered split jobs (janitor requeue / PEL reclaim)
    # must not re-insert shard rows. Shards PK is (doc_id, idx) — probe with
    # the composite key. If shards exist this doc is already split — just
    # advance UPLOADED → PARSING so the state machine converges.
    from core.db.models import Shard

    already = session.get(Shard, (doc_id, 0))
    if already is not None:
        if doc.state == DocState.UPLOADED:
            repo.set_doc_state(session, doc_id, DocState.PARSING)
        return

    local_path = Path(tempfile.mkdtemp(prefix="split-")) / "source.pdf"
    bucket = settings.s3_bucket_raw
    key = s3.raw_key(str(doc_id))
    try:
        s3.download_to(s3.make_s3(settings), bucket, key, str(local_path))
    except Exception as exc:  # pragma: no cover
        raise PlatformError(ErrorCode.PDF_CORRUPT, f"source missing: {exc}") from exc

    import pikepdf

    try:
        with pikepdf.open(str(local_path)) as pdf:
            n_pages = len(pdf.pages)
            outline = extract_outline(str(local_path))
    except pikepdf.PasswordError as exc:
        raise PlatformError(ErrorCode.PDF_ENCRYPTED, "source PDF is encrypted") from exc
    except Exception as exc:
        raise PlatformError(ErrorCode.PDF_CORRUPT, f"cannot open PDF: {exc}") from exc

    if outline:
        bounds = chapter_aligned_bounds(outline, n_pages, settings)
    else:
        bounds = fixed_bounds(n_pages, settings.shard_pages)

    repo.set_doc_state(session, doc_id, DocState.PARSING)
    session.flush()
    doc = repo.get_document(session, doc_id)
    doc.total_shards = len(bounds)
    doc.page_count = n_pages
    repo.insert_shards(session, doc_id, [(b.page_start, b.page_end) for b in bounds])
    session.flush()

    for b in bounds:
        streams.xadd_job(
            redis,
            streams.STREAM_PARSE,
            contracts.ParseJob(
                doc_id=doc_id,
                idx=b.idx,
                page_start=b.page_start,
                page_end=b.page_end,
                source_uri=doc.source_uri,
            ),
        )


def main() -> None:  # pragma: no cover - process entry
    import logging

    from core.db.session import make_engine, make_session_factory
    from core.observability.logging import configure_logging

    settings = get_settings()
    configure_logging(settings.log_level)
    logging.getLogger(__name__).info("splitter worker starting")

    factory = make_session_factory(make_engine(settings))
    redis = streams.make_redis(settings)

    def handler(session, job):
        handle_split(session, job, redis)

    from workers.runner import record_job_error, run_consumer

    run_consumer(
        stream=streams.STREAM_SPLIT,
        consumer=streams.new_consumer_name("splitter"),
        handler=handler,
        session_factory=factory,
        redis=redis,
        prefetch=settings.worker_prefetch,
        on_error=lambda s, j, e: record_job_error(s, j, e, stage="split"),
    )


if __name__ == "__main__":
    main()

File: internal/service/splitter.go
// Package service: the four worker handlers — splitter, parser, embedder —
// wiring internal/pipeline onto internal/store and internal/queue. This is
// the Go counterpart of services/workers.
package service

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"

	"github.com/jackc/pgx/v5"

	"github.com/saufi-opi/lybrix/internal/config"

	"github.com/saufi-opi/lybrix/internal/errors"
	"github.com/saufi-opi/lybrix/internal/objectstore"
	"github.com/saufi-opi/lybrix/internal/pipeline"
	"github.com/saufi-opi/lybrix/internal/queue"
	"github.com/saufi-opi/lybrix/internal/store"
)

// Deps is the shared worker wiring.
type Deps struct {
	Settings *config.Settings
	DB       *store.DB
	Redis    queue.Cmdable2
	S3       *objectstore.Client
	Parser   *pipeline.TwoTierParser
}

// HandleSplit is the splitter handler (PRD §6.2): cheap, CPU-only, never
// touches a heavy parser. Downloads the source once, extracts the page
// count + outline, writes shard rows, and fans out one doc.parse job per
// shard.
func HandleSplit(ctx context.Context, deps Deps, tx pgx.Tx, job map[string]any) error {
	s := deps.Settings
	docID, err := jobString(job, "doc_id")
	if err != nil {
		return err
	}
	doc, err := deps.DB.GetDocument(ctx, docID)
	if err != nil {
		return err
	}
	if doc == nil {
		return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("document %s vanished", docID))
	}

	// Idempotency: re-delivered split jobs (janitor requeue / PEL reclaim)
	// must not re-insert shard rows. Shards PK is (doc_id, idx) — probe with
	// the composite key. If shards exist this doc is already split — just
	// advance UPLOADED → PARSING so the state machine converges.
	already, err := deps.DB.GetShard(ctx, docID, 0)
	if err != nil {
		return err
	}
	if already != nil {
		if doc.State == store.StateUploaded {
			return deps.DB.SetDocState(ctx, tx, docID, store.StateParsing, nil, nil)
		}
		return nil
	}

	tmpDir, err := os.MkdirTemp("", "split-")
	if err != nil {
		return err
	}
	defer os.RemoveAll(tmpDir)
	localPath := filepath.Join(tmpDir, "source.pdf")
	bucket := s.S3BucketRaw
	key := objectstore.RawKey(docID)
	if err := deps.S3.DownloadTo(ctx, bucket, key, localPath); err != nil {
		return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("source missing: %v", err))
	}

	// EPUB branch (PLAN.md item 19): pdfcpu would raise PDF_CORRUPT on a
	// zip container, so the EPUB path skips PageCount/ExtractBookmarks
	// entirely and emits exactly ONE synthetic shard (idx=0, pages 1-1) —
	// the parser routes it straight to docling-serve, which parses EPUB
	// natively, and treats the whole book as one "page".
	if doc.MimeType != nil && *doc.MimeType == "application/epub+zip" {
		return splitEpubSingleShard(ctx, deps, tx, doc, tmpDir)
	}

	pageCount, err := pipeline.PageCount(localPath)
	if err != nil {
		if code := pipeline.ClassifyPDFError(err); code == "PDF_ENCRYPTED" {
			return errors.NewPlatformError(errors.CodePDFEncrypted, "source PDF is encrypted")
		}
		return errors.NewPlatformError(errors.CodePDFCorrupt, fmt.Sprintf("cannot open PDF: %v", err))
	}
	outline := ExtractBookmarks(localPath)

	var bounds []pipeline.ShardBound
	if len(outline) > 0 {
		bounds, err = pipeline.ChapterAlignedBounds(outline, pageCount, s.ShardPages)
	} else {
		bounds, err = pipeline.FixedBounds(pageCount, s.ShardPages, 1)
	}
	if err != nil {
		return err
	}

	if err := deps.DB.SetDocState(ctx, tx, docID, store.StateParsing, nil, nil); err != nil {
		return err
	}
	totalShards := len(bounds)
	if err := updateSplitMeta(ctx, tx, docID, totalShards, pageCount); err != nil {
		return err
	}
	if _, err := deps.DB.InsertShards(ctx, tx, docID, toBoundPairs(bounds), 0); err != nil {
		return err
	}
	for _, b := range bounds {
		job := queue.ParseJob{
			SchemaVersion: queue.SchemaVersion,
			DocID:         docID,
			Idx:           b.Idx,
			PageStart:     b.PageStart,
			PageEnd:       b.PageEnd,
			SourceURI:     doc.SourceURI,
		}
		if _, err := queue.XAddJob(ctx, deps.Redis, queue.StreamParse, job); err != nil {
			return err
		}
	}
	slog.Info("split complete", "doc", docID, "pages", pageCount, "shards", totalShards)
	return nil
}

// splitEpubSingleShard is the EPUB splitter tail: advance UPLOADED →
// PARSING, write one synthetic shard (idx=0, page_start=1, page_end=1),
// enqueue its ParseJob — identical to the PDF path's tail but with the
// pdfcpu/bookmark steps skipped.
func splitEpubSingleShard(ctx context.Context, deps Deps, tx pgx.Tx, doc *store.Document, tmpDir string) error {
	docID := doc.ID
	// Idempotency mirrors the PDF path's GetShard probe above.
	already, err := deps.DB.GetShard(ctx, docID, 0)
	if err != nil {
		return err
	}
	if already != nil {
		if doc.State == store.StateUploaded {
			return deps.DB.SetDocState(ctx, tx, docID, store.StateParsing, nil, nil)
		}
		return nil
	}
	if err := deps.DB.SetDocState(ctx, tx, docID, store.StateParsing, nil, nil); err != nil {
		return err
	}
	if err := updateSplitMeta(ctx, tx, docID, 1, 1); err != nil {
		return err
	}
	if _, err := deps.DB.InsertShards(ctx, tx, docID, [][2]int{{1, 1}}, 0); err != nil {
		return err
	}
	job := queue.ParseJob{
		SchemaVersion: queue.SchemaVersion,
		DocID:         docID,
		Idx:           0,
		PageStart:     1,
		PageEnd:       1,
		SourceURI:     doc.SourceURI,
	}
	if _, err := queue.XAddJob(ctx, deps.Redis, queue.StreamParse, job); err != nil {
		return err
	}
	slog.Info("split complete (epub, single shard)", "doc", docID)
	return nil
}

// updateSplitMeta sets total_shards + page_count in one statement.
func updateSplitMeta(ctx context.Context, tx pgx.Tx, docID string, totalShards, pageCount int) error {
	_, err := tx.Exec(ctx,
		`UPDATE documents SET total_shards = $2, page_count = $3, updated_at = NOW() WHERE id = $1`,
		docID, totalShards, pageCount)
	return err
}

func toBoundPairs(bounds []pipeline.ShardBound) [][2]int {
	out := make([][2]int, len(bounds))
	for i, b := range bounds {
		out[i] = [2]int{b.PageStart, b.PageEnd}
	}
	return out
}

func jobString(job map[string]any, key string) (string, error) {
	v, ok := job[key].(string)
	if !ok || v == "" {
		return "", fmt.Errorf("job field %q missing", key)
	}
	return v, nil
}
